### PR TITLE
Intrepid2 hierarchical wedges

### PR DESCRIPTION
**Pull request by CamelliaDPG**
@trilinos/intrepid2

## Motivation
This PR adds bases for H(grad),  H(curl), H(div), and H(vol) on the wedge.  This includes both high-order nodal wedges and high-order hierarchical wedges; previously only first and second order nodal wedges were available for H(grad), and only first order were available for H(curl) and H(div) -- no H(vol) basis was available.  This follows #10662, which provided hierarchical bases for the triangle, and #11009, which did so for the tetrahedron.

A planned forthcoming PR will similarly add hierarchical bases on the pyramid.

## Testing
Some basic "basis equivalence" tests, which establish that the new bases span the same space as existing low-order nodal bases, are included here.  Basis equivalence tests are also included to verify that the new high-order nodal and high-order hierarchical bases span the same space.  Additionally, I have done offline comparison with the ESEAS implementation of each basis, testing up to 8th order, with excellent agreement.

### CODE REVIEW

**Comment by CamelliaDPG** [comment]
Latest test failure appears to be due to Issue #10999 -- I'm guessing that this failure does not happen always, as #11009, which was also affected, did eventually go through.  Setting RETEST to see if we can get this through.

File: packages/intrepid2/src/Discretization/Basis/Intrepid2_DerivedBasis_HVOL_WEDGE.hpp [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+    /** \brief  Constructor.
+        \param [in] polyOrder - the polynomial order to use in both dimensions.
+        \param [in] pointType - type of lattice used for creating the DoF coordinates.
+     */
+    Basis_Derived_HVOL_WEDGE(int polyOrder, const EPointType pointType=POINTTYPE_DEFAULT) : Basis_Derived_HVOL_WEDGE(polyOrder,polyOrder,pointType) {}
+
+
+    /** \brief True if orientation is required
+    */
+    virtual bool requireOrientation() const override {
```
Comment (mperego) [inline packages/intrepid2/src/Discretization/Basis/Intrepid2_DerivedBasis_HVOL_WEDGE.hpp:112]: HVol orientation is always false.

**Review by mperego — COMMENTED** [review]
@CamelliaDPG Looks good! I have a couple of comments.

File: packages/intrepid2/src/Discretization/Basis/Intrepid2_DerivedBasis_HDIV_WEDGE.hpp [hunk 2]
```diff
@@ -0,0 +1,420 @@
...
+        \param [in] polyOrder - the polynomial order to use in all dimensions.
+        \param [in] pointType - type of lattice used for creating the DoF coordinates.
+     */
+    Basis_Derived_HDIV_WEDGE(int polyOrder, const EPointType pointType=POINTTYPE_DEFAULT) : Basis_Derived_HDIV_WEDGE(polyOrder, polyOrder, pointType) {}
+
+    /** \brief True if orientation is required
+    */
+    virtual bool requireOrientation() const override
+    {
+      return (this->getDofCount(1,0) > 0); //if it has edge DOFs, than it needs orientations
```
[the remark this answers is not in the document]
Comment (CamelliaDPG) [reply packages/intrepid2/src/Discretization/Basis/Intrepid2_DerivedBasis_HDIV_WEDGE.hpp]: I've fixed this here and in the QUAD and HEX bases.

File: packages/intrepid2/src/Discretization/Basis/Intrepid2_DerivedBasis_HCURL_WEDGE.hpp [hunk 3]
```diff
@@ -0,0 +1,466 @@
...
+    
+    /** \brief  Constructor.
+        \param [in] polyOrder - the polynomial order to use in all dimensions.
+        \param [in] pointType - type of lattice used for creating the DoF coordinates.
+     */
+    Basis_Derived_HCURL_WEDGE(int polyOrder, const EPointType pointType=POINTTYPE_DEFAULT) : Basis_Derived_HCURL_WEDGE(polyOrder, polyOrder, pointType) {}
+
+    /** \brief True if orientation is required
+    */
+    virtual bool requireOrientation() const override
```
[the remark this answers is not in the document]
Comment (CamelliaDPG) [reply packages/intrepid2/src/Discretization/Basis/Intrepid2_DerivedBasis_HCURL_WEDGE.hpp:436]: It's a quick fix; might as well do it now!  I've fixed this here and in the QUAD and HEX bases.

File: packages/intrepid2/src/Discretization/Basis/Intrepid2_DerivedBasis_HVOL_WEDGE.hpp (hunk 1 above)
Comment (CamelliaDPG) [reply packages/intrepid2/src/Discretization/Basis/Intrepid2_DerivedBasis_HVOL_WEDGE.hpp:112]: Also now fixed -- looks like this one was the only instance of this issue (the other derived H(vol) bases already returned false).

**Comment by CamelliaDPG** [comment]
> @CamelliaDPG Looks good! I have a couple of comments.

Thanks, @mperego!

**Comment by CamelliaDPG** [comment]
@mperego, would you please approve this when you get a minute?

**Review by mperego — APPROVED** [review]
Thanks @CamelliaDPG for addressing my comments!

**Pull request merged**